R/clean_data_helpers.R

```diff
@@ -139,26 +139,50 @@ scan_in_character <- function(x, x_name) {
       )
     )
   )
-
   date_count <- sum(!is.na(are_date))
   character_count <- character_count - date_count
 
   # convert everything to numeric and get the numeric count
   are_numeric <- suppressWarnings(as.numeric(x))
   numeric_count <- sum(!is.na(are_numeric))
 
+  # get the count of ambiguous values
+  # ambiguous here are date values which after conversion to numeric turns out
+  # to be numeric as well
+  are_numeric_and_date <- !is.na(are_date) & !is.na(are_numeric)
   numeric_and_date_count <- sum(!is.na(are_date) & !is.na(are_numeric))
 
+  # subtracting the numeric_and_date_count here as they have already been
+  # counted in the date count
+  character_count <- character_count - (numeric_count - numeric_and_date_count)
+
+  # convert the numeric values into date to detect potential date values within
+  # the numeric. If some are date, get the second count of ambiguous and date
+  # values
+  oldest_date <- Sys.Date() - lubridate::years(50) # Set the first date to 50
+                                                  #years before the current date
+  date_values <- lubridate::as_date(are_numeric)
+
+  # do not account for the previous ambiguous values that have already been
+  # counted
+  date_values[are_numeric_and_date] <- NA
+  valid_dates <- date_values >= oldest_date & date_values <= Sys.Date()
+
+  # converting non-date values to NA to prevent from error in the intersection
+  # below
+  valid_dates[!valid_dates] <- NA
+  numeric_and_date_count <- numeric_and_date_count +
+    sum(!is.na(valid_dates) & !is.na(are_numeric))
+  date_count <- date_count + sum(valid_dates, na.rm = TRUE)
```

Comment:
I believe this change is going back to the old logic. The idea in this PR is that we don't have to try and identify dates from numerics specifically since we already identified them.

In other words, the proposed logic is:

- count dates
- count numerics
- DO NOT count ambiguous strings. They are identified as an emerging property at the intersection of dates & numerics.

Reply:
This is the logic that I understood from our previous discussions:
1. convert to date and get the first date count. When there are numerics among these dates, get the first count of numeric.
2. convert the remaining non-date into numeric and get the second numeric count. Then check whether there are potential dates among these numerics. If yes, get the second date count.
3. get the logical count

Take the following as an example:

```{r}
x = c("16372", "Jun /23/Y", "16384", "16509", "Jun /30/Y", "16382", "16272", "16376", "02/09/2015", "16444", "16503", "16686", "16370", "16347", "16440")
```

* after conversion to date, only `02/09/2015` will be counted, and this is not convertable to numeric. At this point:
date_count = 1
numeric_count = 0
character_count = 14
logical_count = 0

* after conversion to numeric, the followings will be counted: `16372, 16384, 16509, 16382, 16272, 16376, 16444, 16503, 16686, 16370, 16347, 16440`. These 12 elements turn out to also be correct date values when the `lubridate::as_date()` is applied to them. Hence increasing the count of date values. At this point:
date_count = 13 (1 + 12)
numeric_count = 12
character_count = 2 ("Jun /23/Y", "Jun /30/Y")
logical_count = 0

When I apply the logic you described above, I get the following result:
date_count = 1
numeric_count = 12
character_count = 2
logical_count = 0

This means you are no longer accounting for numeric values that are date at the same time. I understand that this prevents from guessing whether a value is a date or not, but it is important specially if we take into account for the fact that this is a frequent scenario when data is imported from MS Excel.

Reply:
What about strings like `"20220101"`? Should they be reported as ambiguous strings? They will be reported as numerics & dates in all the solutions proposed here

Reply:
I see what you mean now. I then propose to use the `date_guess()` function which is used in other places of the code.

This:

- simplifies the logic since we don't have to do multiple rounds of date conversion
- ensures consistency in the results users will get in `scan_data()` and `standardize_dates()`

How does it sound?

If I understood correctly though, in the specific example you shared above, dates should still not be counted as they are outside the allowed range, aren't they?

Reply:
The date guesser in `standardize_dates()` returns wrong dates (returns years around 1944, which is outside of 50 years back from today's date) when applied to the example data above.

The date guessers used in `scan_data()` (`lubridate::parse_date_time()` and `lubridate::as_date()`) provide correct dates (dates on that column are expected to be around years 2015 and 2014).

What you are proposing here highlights the problem with the date guesser in `standardize_dates()` function. This needs to be fixed at the same time with the current problem or before.

As we already have an issue about `date_guess()`, I propose the following:

* address them in different PRs, starting by fixing `date_guess()` (using the approach in this implementation)
* then use the `date_guess()` within the `scan_data()` function as you suggest.

Reply:
> As we already have an issue about date_guess(), I propose the following:
>
> - address them in different PRs, starting by fixing date_guess() (using the approach in this implementation)
> - then use the date_guess() within the scan_data() function as you suggest.

Yes, that seems like a sensible strategy. Let's do this and put this PR on hold for now. :+1: